src/runtime/hexagon_remote/remote.cpp

```diff
@@ -44,15 +95,239 @@ int halide_do_task(void *user_context, halide_task_t f, int idx,
                    uint8_t *closure) {
     return f(user_context, idx, closure);
 }
+void lock_work_queue() {
+    qurt_mutex_lock(&(work_queue.work_mutex));
+}
+void unlock_work_queue() {
+    qurt_mutex_unlock(&(work_queue.work_mutex));
+}
+void goto_work(work *owned_job) {
+    work myjob;
+    // You can work only if you get a lock on the work queue.
+    unsigned int tid = qurt_thread_get_id();
+    FARF(LOW, "HVX_TP: %d: goto_work: Trying to get a lock on the work queue\n", tid);
+
+    // ***********************
+    // *** Lock work queue ***
+    // ***********************
+    lock_work_queue();
+    FARF(LOW, "HVX_TP: %d: goto_work: got a lock on the work queue\n", tid);
+
+    // If I'm a job owner, then I was the thread that called
+    // do_par_for, and I should only stay in this function while there
+    // are active workers on the job (i.e. running()) or if the job
+    // hasn't been claimed entirely (!claimed()). If I'm a lowly worker
+    // thread, I should stay in this function as long as the work queue is running.
+    while (owned_job != NULL ? (owned_job->running() || !owned_job->claimed())
+           : work_queue.running()) {
+
+        FARF(LOW, "HVX_TP: %d: goto_work: In the main goto_work loop \n", tid);
+        int num_jobs = work_queue.num_jobs, job_idx;
+        work *jobs = work_queue.jobs, *orig_job;
+        bool job_found = false;
+        if (!owned_job) {
```

Comment:
I think the control flow here could be simplified if it looked something like this:

    work *job_to_do = NULL;  // instead of orig_job
    if (!owned_job) {
        // This is a worker thread, try to find a job to to do.
        job_to_do = find_nonempty_job(...);
        if (!job_to_do) {
            // go to sleep
        }
    } else {
        // This is the main thread
        if (owned_job->claimed()) {
            // Go to sleep, wait for the worker thread to wake us up.
        } else {
            // Do some work.
            job_to_do = owned_job;
        }
    }
    // Do some work on job_to_do.

After writing that out, it didn't make as big of a difference as I hoped. I just don't like the redundancy in the "job exists" state between job_found and orig_job.